src/GameState.cpp

```diff
@@ -320,6 +320,48 @@ std::optional<std::optional<Color>> GameState::getOutcome() const {
     // TODO:
     // If we're in check, see if there's a way to get out. (If not, our opponent wins.)
     // If we're not in check, see if there's any move that doesn't put us in check. (If not, it's stalemate.)
+
+
+    // Check for checkmate if king is in check
+    if (this->isCheck(this->getTurn()))
+    {
+        int legalMoveFlag = 0;
+        for (const auto [square, piece] : pieceMap) {
+            // Iterating over the pieces of the same color and checking for legal moves
+            if (piece.color != this->turn) continue;
+
+            for (auto move : this->generateMoves(square)) {
+                if(isLegal(move))
+                    legalMoveFlag++;
+                }
+            if(legalMoveFlag == 0)
+            {
+                return this->turn == Color::White? Chess::Color::Black : Color::White;
+            }
+        }
+    }
+
+//    Check for stalemate
+    else
+    {
+        int legalMoveFlag = 0;
+        for (const auto [square, piece] : pieceMap) {
```

Comment:
This has much the same structure as the non-stalemate loop above.

We could instead _first_ check if the current player has legal moves. If so, the game is not over, and we're done. Otherwise, the game is definitely over, and then we just have to determine how (checkmate or stalemate?) by seeing if the player is in check. This is logically equivalent to the current code, but we only need to express the check for legal moves once.

(This is the transpose of what I suggested in the "TODO" above. Sorry about that - I missed the commonality of checking for legal moves in both checkmate and stalemate until seeing it laid out in code.)